lib/Model/DlibCnnHogModel/DlibCnnHogModel.php

```diff
@@ -0,0 +1,292 @@
+<?php
+/**
+ * @copyright Copyright (c) 2020, Matias De lellis <mati86dl@gmail.com>
+ *
+ * @author Matias De lellis <mati86dl@gmail.com>
+ *
+ * @license GNU AGPL version 3 or any later version
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OCA\FaceRecognition\Model\DlibCnnHogModel;
+
+use OCP\IDBConnection;
+
+use OCA\FaceRecognition\Helper\MemoryLimits;
+
+use OCA\FaceRecognition\Service\FileService;
+use OCA\FaceRecognition\Service\ModelService;
+use OCA\FaceRecognition\Service\SettingsService;
+
+use OCA\FaceRecognition\Model\IModel;
+
+class DlibCnnHogModel implements IModel {
```

Comment:
If I am not mistaken, this model is composition of two other models, so maybe it is better to have reference to them in this class (`private DlibHogModel $hogModel; private DlibCnnModel $cnnModel;`). You would not keep `private $cfd;` here (or others). Then, you would proxy all method to them. For example, `isInstalled` would return something like `$hogModel->isInstalled() && $cnnModel->isInstalled()`. `install`, `meetDependencies`, `getMaximumArea` etc similar logic. You would also forward call for `detectLandmarks` and `computeDescriptor` to `cnnModel`.

It was long time ago since I was learning design pattern:), but essentially I am describing https://en.wikipedia.org/wiki/Decorator_pattern

Reply:
Ohh .. you're right .. I will try to do it later..

Reply:
Well .. It is not exactly what you expected, but I think for now it is enough ..

Reply:
No, this is EXACTLY what I had in mind, every line of it:) Pure beauty now, IMHO:)